schematools/events/__init__.py

```diff
@@ -16,41 +17,313 @@
 
 logger = logging.getLogger(__name__)
 
 
+class UnknownRelationException(Exception):
+    """Custom Exception to signal the fact that an incoming
+    event is unsing an unknown type of relation for
+    which we do not have configuration info available.
+    """
+
+    pass
+
+
 def fetch_insert_data(event_data):
-    return event_data["entity"]
+    # remove dict values from the event. We only handle scalar datatypes.
+    # Json blobs in the events are containing irrelevant GOB-only data.
+    return {k: v for k, v in event_data["entity"].items() if not isinstance(v, dict)}
 
 
 def fetch_update_data(event_data):
+    # Update data has a different structure in the events. We convert it
+    # into a single dict (like for insert data), so we can handle
+    # the data in the exact same way.
     update_data = {}
     for modification in event_data["modifications"]:
-        # XXX skip geometrie for now, has geojson format (no wkt)
+        # XXX skip geometrie for now, has geojson format (should be wkt)
         if modification["key"] == "geometrie":
             continue
         update_data[modification["key"]] = modification["new_value"]
     return update_data
 
 
-EVENT_TYPE_MAPPING = {
-    "ADD": ("insert", False, fetch_insert_data),
-    "MODIFY": ("update", True, fetch_update_data),
-    "DELETE": ("delete", True, None),
+FK_TABLE = False
+NM_TABLE = True  # nm_table_id is not None
+
+
+# Configuration information to map the event-type
+# to the following fields (package into a DbInfo dataclass):
+# - db_operation_name (INSERT/UPDATE)
+# - needs_select: generated SQL needs to select a record
+# - needs_values: generated SQL needs values (to insert/update)
+# - data_fetcher: function to be used to extract data from the event
+
+EVENT_TYPE_MAPPINGS = {
+    FK_TABLE: {
+        "ADD": ("update", True, True, fetch_insert_data),
+        "MODIFY": ("update", True, True, fetch_update_data),
+        "DELETE": ("update", True, True, None),
+    },
+    NM_TABLE: {
+        "ADD": ("insert", False, True, fetch_insert_data),
+        "MODIFY": ("update", True, True, fetch_update_data),
+        "DELETE": ("delete", True, False, None),
+    },
 }
 
+
+@dataclass
+class DbInfo:
+    db_operation_name: str
+    needs_select: bool
+    needs_values: bool
+    data_fetcher: Optional[Callable[[Dict], Dict]]
+
+    @classmethod
+    def from_event_type(cls, is_nm_table: bool, event_type: str):
+        return cls(*EVENT_TYPE_MAPPINGS[is_nm_table][event_type])
+
+
+# Translation table of GOB relation names to amsterdam schema configuration
+# Events for relations contain a 'rel' field that is an internal GOB name
+# We need to associate configuration info with these events to be able
+# to map the event to the correct tables according to the amsterdam schema.
+
+# Structure
+# <gob rel name> :
+#       (<amschema dataset>, <amschema table>, <nm relation table>, <amschema fieldname>)
+
 COLLECTION_TO_SCHEMA = {
```

Comment:
Is deze mapping alleen voor gebieden noodzakelijk?

Reply:
Dit zal voor alle GOB dataset geconfigureerd moeten worden (zodra dat bekend is). Helaas moeten we met deze internals van de GOB systemen werken. Ideaal zou zijn dat de events helemaal self-contained zijn en in termen van het amsterdam schema zijn gedefinieerd.